### ISSUE TITLE
QEMU: Intermittent connection failure between docker containers

### ISSUE THREAD

**Issue by geissonator** [openbmc/openbmc#3101]
I see this maybe 5% of the time in our QEMU job.  https://openpower.xyz/job/openbmc-test-qemu-ci/

The script gets the first docker container (QEMU running witherspoon) booted up and logged into and then boots up the other docker container to run the robot tests against it.  All of those robot tests fail to connect and talk to the first docker container.

First docker container getting logged into
```
10:00:24 root@witherspoon:~# OPENBMC-READY
'
10:00:24 + grep -q OPENBMC-READY
10:00:24 + echo 'QEMU is ready!'
10:00:24 QEMU is ready!
```

Start up robot test docker container:
```
10:00:28 + docker run --rm --user root --env HOME=/var/lib/jenkins-slave --env IP_ADDR=172.17.0.2 --env SSH_PORT=22 --env HTTPS_PORT=443 --env MACHINE=witherspoon --workdir /var/lib/jenkins-slave --volume /var/lib/jenkins-slave/workspace/openbmc-test-qemu-ci:/var/lib/jenkins-slave --tty openbmc/ubuntu-robot-qemu /var/lib/jenkins-slave/run-robot.sh
10:00:29 + MACHINE=witherspoon
10:00:29 ++ whoami
10:00:29 + ROBOT_CODE_HOME=/tmp/root/11454/obmc-robot/
10:00:29 + ROBOT_TEST_CMD='tox -e witherspoon -- --argumentfile test_lists/QEMU_CI tests'
10:00:29 + git clone https://github.com/openbmc/openbmc-test-automation.git /tmp/root/11454/obmc-robot/
```

Everything fails in test:
```
10:00:29 + export OPENBMC_HOST=172.17.0.2
10:00:29 + OPENBMC_HOST=172.17.0.2
10:00:29 + export SSH_PORT=22
10:00:29 + SSH_PORT=22
10:00:29 + export HTTPS_PORT=443
10:00:29 + HTTPS_PORT=443
10:00:29 ++ tox -e witherspoon -- --argumentfile test_lists/QEMU_CI tests
10:01:27 + 'witherspoon create: /tmp/root/11454/obmc-robot/.tox/witherspoon
10:01:27 witherspoon installdeps: -r/tmp/root/11454/obmc-robot/requirements.txt
10:01:27 witherspoon installed: The directory '\''/var/lib/jenkins-slave/.cache/pip/http'\'' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo'\''s -H flag.,asn1crypto==0.24.0,certifi==2018.4.16,cffi==1.11.5,chardet==3.0.4,cryptography==2.2.2,enum34==1.1.6,idna==2.6,ipaddress==1.0.22,paramiko==2.0.2,pyasn1==0.4.2,pycparser==2.18,requests==2.18.4,robotframework==3.0.3,robotframework-requests==0.4.7,robotframework-scplibrary==1.1.1,robotframework-sshlibrary==2.1.3,scp==0.10.2,six==1.11.0,urllib3==1.22
10:01:27 witherspoon runtests: PYTHONHASHSEED='\''4030531424'\''
10:01:27 witherspoon runtests: commands[0] | -mkdir -p /tmp/root/11454/obmc-robot/.tox/witherspoon/tmp
10:01:27 witherspoon runtests: commands[1] | bash /tmp/root/11454/obmc-robot/tools/generate_argumentfile.sh
10:01:27 witherspoon runtests: commands[2] | cp bin/ssh_pw /tmp/root/11454/obmc-robot/.tox/witherspoon/bin/ssh_pw
10:01:27 witherspoon runtests: commands[3] | python -m robot.run --argumentfile /tmp/root/11454/obmc-robot/.tox/witherspoon/tmp/argument_file.txt --argumentfile test_lists/QEMU_CI tests
10:01:27 ==============================================================================
10:01:27 Tests :: Set metadata for test suite.                                         
10:01:27 ==============================================================================
10:01:27 Tests.Security                                                                
10:01:27 ==============================================================================
10:01:27 Tests.Security.Test Ssl :: This testsuite is for testing SSL connection to ...
10:01:27 ==============================================================================
10:01:27 Test SSL Connection :: This testcase is for testing the SSL connec... | FAIL |
10:01:27 Parent suite setup failed:
10:01:27 SSHException: Error reading SSH protocol banner
...
```

**Comment by geissonator** [comment]
Hit over the weekend: https://openpower.xyz/job/openbmc-test-qemu-ci/2057/

**Comment by geissonator** [comment]
Some FFDC was able to be collected from the docker container running QEMU when in the fail state.

https://openpower.xyz/job/openbmc-test-qemu-ci/2642/artifact/logs/20180606103129344319_TestsSecurityTestSsl/20180606103129344319_TestSSLConnection/20180606103129344319_BMC_journalctl_nopager.txt

The main issue is this, causing the REST server to not be able to start
```
Jun 06 10:26:13 witherspoon sync_inventory_items.py[1108]: Traceback (most recent call last):
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: xz decompression failed, data probably corrupt
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: squashfs_read_data failed to read block 0xafc844
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [afc844]
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read page, block afc844, size 8e20
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [afc844]
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read page, block afc844, size 8e20
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [afc844]
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read page, block afc844, size 8e20
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [afc844]
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read page, block afc844, size 8e20
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [afc844]
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read page, block afc844, size 8e20
Jun 06 10:26:14 witherspoon kernel[978]: [  103.543621] SQUASHFS error: xz decompression failed, data probably corrupt
Jun 06 10:26:14 witherspoon kernel[978]: [  103.545754] SQUASHFS error: squashfs_read_data failed to read block 0xafc844
Jun 06 10:26:14 witherspoon sync_inventory_items.py[1108]:   File "/usr/sbin/sync_inventory_items.py", line 23, in <module>
Jun 06 10:26:14 witherspoon sync_inventory_items.py[1108]:     import subprocess
Jun 06 10:26:14 witherspoon sync_inventory_items.py[1108]:   File "/usr/lib/python2.7/subprocess.py", line 74, in <module>
Jun 06 10:26:14 witherspoon sync_inventory_items.py[1108]:     import fcntl
Jun 06 10:26:14 witherspoon sync_inventory_items.py[1108]: ImportError: /usr/lib/python2.7/lib-dynload/fcntl.so: cannot read file data: Input/output error
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [afc844]
Jun 06 10:26:14 witherspoon kernel: SQUASHFS error: Unable to read page, block afc844, size 8e20
```

**Comment by amboar** [comment]
Maybe we should apply https://github.com/amboar/linux/commit/4620d829f7615dadbca793bd24818f00f16efbc0 - this might help get to the bottom of what happened.

**Comment by charleshofer** [comment]
So, I threw together an image with @amboar 's change and reran our test suite on it until we hit the error. Looks something like

```
Jun 11 16:43:53 witherspoon phosphor-gevent[1226]:   File "/usr/sbin/phosphor-gevent", line 47, in <module>
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: xz decompression failed, data probably corrupt: -5
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: squashfs_read_data failed to read block 0xa6b8b8
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [a6b8b8]
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: Unable to read page, block a6b8b8, size 4d58
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [a6b8b8]
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: Unable to read page, block a6b8b8, size 4d58
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: Unable to read fragment cache entry [a6b8b8]
Jun 11 16:43:53 witherspoon kernel: SQUASHFS error: Unable to read page, block a6b8b8, size 4d58
```

I think that -5 maps to an EIO, input/output error that's used for physical read/write problems. Doesn't mean much to me, but maybe @amboar knows more about it?

**Comment by amboar** [comment]
That `EIO` would be from https://github.com/openbmc/linux/blob/dev-4.13/fs/squashfs/xz_wrapper.c#L182 which means we've hit https://github.com/openbmc/linux/blob/dev-4.13/fs/squashfs/xz_wrapper.c#L173 . I guess we want to log `xz_err` as well.

**Comment by charleshofer** [comment]
Yeah @amboar, if you think that'd help us get to the bottom of this. If you have a change that will log xz_err, I could get it into a build and run QEMU tests until we hit the issue again.

**Comment by amboar** [comment]
@charleshofer Try https://github.com/amboar/linux/commit/53313fe68684cd857451479298e6965512312e16

**Comment by geissonator** [comment]
Just to reiterate, I don't have QEMU CI enabled in a way that it will log any type of error to a commit when it fails because of this issue.  We're seeing multiple changes make it onto real hardware that are breaking the hardware due to this.

**Comment by charleshofer** [comment]
I'll make this a higher priority. I'm not sure if I'll be able to put an image together with @amboar's change without having him rebase onto the latest kernel though?

**Comment by amboar** [comment]
@charleshofer @geissonator I've rebased the changes on dev-4.17 and pushed. They are here:

https://github.com/openbmc/linux/compare/dev-4.17...amboar:squashfs-decompress-failure

The rebase worked without issue, so the patches should have applied naturally to a rebased openbmc tree that pulled in the new kernel.

**Comment by charleshofer** [comment]
I have an image with your changes in it @amboar. I was able to run our CI tests on it, hit the QEMU issue, and get FFDC back. I have the journal here:
[journal.txt](https://github.com/openbmc/openbmc/files/2189353/journal.txt)
The important thing is probably

```

Jul 02 20:15:09 witherspoon kernel: SQUASHFS error: squashfs_xz_uncompress: xz decompressor failed: 7
Jul 02 20:15:09 witherspoon kernel: SQUASHFS error: xz decompression failed, data probably corrupt: -5

```

**Comment by amboar** [comment]
7 is `XZ_DATA_ERROR`. So it looks like it might be a memory corruption problem? We're having those elsewhere - it tends to hit in the network a bit. You can see we have a bunch of issues open here: https://github.com/openbmc/qemu/issues

**Comment by charleshofer** [comment]
@amboar Any idea on how to fix the memory corruption issue that's causing this? If we can't, maybe we can set up something within Jenkins so that we just retry the tests if we hit it.

**Comment by charleshofer** [comment]
I'm working on a [change](https://gerrit.openbmc-project.xyz/#/c/openbmc/openbmc-build-scripts/+/11625/) to restart QEMU if we see this error in the log.

**Comment by amboar** [comment]
I've just seen the following the one local qemu session running a witherspoon image:

```
[   89.506337] SQUASHFS error: xz decompression failed, data probably corrupt
[   89.507663] SQUASHFS error: squashfs_read_data failed to read block 0x3e6440
[   89.508059] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[   89.508411] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[   89.508826] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[   89.509167] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[   89.509465] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[   89.509733] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[   89.510049] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[   89.510316] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[   89.510684] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[   89.510949] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[  113.120192] SQUASHFS error: xz decompression failed, data probably corrupt
[  113.121679] SQUASHFS error: squashfs_read_data failed to read block 0x3e6440
[  113.122086] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[  113.122458] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[  138.924616] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[  138.926080] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[  174.103842] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[  174.105332] SQUASHFS error: Unable to read page, block 3e6440, size 5380
[  206.457987] SQUASHFS error: Unable to read fragment cache entry [3e6440]
[  206.458380] SQUASHFS error: Unable to read page, block 3e6440, size 5380
```

**Comment by mdmillerii** [comment]
Have people checked if there squashfs is fitting in the allotted space?  A check during image assembly was only just added.

**Comment by amboar** [comment]
@mdmillerii that thought crossed my mind as well.

**Comment by charleshofer** [comment]
So, I've been trying to test that change that I've been working on to restart QEMU when we hit this bug. https://gerrit.openbmc-project.xyz/#/c/openbmc/openbmc-build-scripts/+/11625/

The problem is testing it. Just about every day the past week and a half, I've sat here running that script with my change, but failing to actually see the QEMU squashfs memory error. The script runs fine if we don't hit the error, but I have no idea if the restart actually works. Is there any chance we could just merge the change and see if it works when hit the QEMU issue in the wild? I can't seem to be able to hit the QEMU error in my testing.

**Comment by charleshofer** [comment]
@amboar It's possible that I couldn't hit the error because of the check. The image I used for testing was the latest one off of AFS. Just to be sure, I could pull one from CI, run the tests a bunch of times, and see if the problem has gone away.

**Comment by charleshofer** [comment]
We also have a Jenkins job for GitHub enterprise and a CT regression test that runs on master every once in a while. Looking back at the builds, we haven't hit this issue since July 24th. So, maybe the squashfs size check fixed it.

**Comment by mdmillerii** [comment]
Have we hit the size check?  I thought it only caused the build to fail and didn't try to remedy the cause.

**Comment by charleshofer** [comment]
Well, the image is built in a separate Jenkins job. If the build fails, we don't even get to the QEMU job. I haven't seen a QEMU job fail since the 24th.

**Comment by Kitsok** [comment]
Hello all!
I'm porting OpenBMC to a custom Grantley motherboard and faced the same issue - squashfs errors during startup. It looks like the flash is removed during the BMC operations.
I've found that if I disable xyz.openbmc_project.FruDevice.service the error disappears.
I have manually started /usr/sbin/fru-device and after very short time of it walking through the I2C buses and devices on the buses I've got squashfs error. 

On my real hardware this is probably caused by the connection of the SPI flash through I2C mux that is affected by the fru-device.
